MQTT: Managing binary payloads

MrMep opened the pull request:
Hello,
background: I wrote a HA camera component that gets the image from a binary payload. I'm testing it with Zanzito (https://play.google.com/store/apps/details?id=it.barbaro.zanzito) and it works apparently well: it gets the image and correctly displays it in the front-end.
But I had to make the changes I'm proposing here: the message was being blocked because the utf-8 decoding failed.
As far as I know, the utf-8 encoding is required for the topic, not for the payload. What I did here was try the utf-8 decoding, but even if unsuccessful, it dispatches the message anyway.
Is there anything else I'm missing?
thanks
Gianluca

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

MrMep committed:
Managing binary payloads

Hello,
background: I wrote a HA camera component that gets the image from a binary payload. I'm testing it with Zanzito (https://play.google.com/store/apps/details?id=it.barbaro.zanzito) and it works apparently well: it gets the image and correctly displays it in the front-end.
But I had to make the changes I'm proposing here: the message was being blocked because the utf-8 decoding failed.
As far as I know, the utf-8 encoding is required for the topic, not for the payload. What I did here was try the utf-8 decoding, but even if unsuccessful, it dispatches the message anyway.
Is there anything else I'm missing?
thanks
Gianluca

homeassistant commented:
Hi @MrMep, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#6976).

Once you do that we will be able to review and accept this pull request.

Thanks!

armills commented:
These are my thoughts:

1) We'd need a new event in hass for binary data, maybe `SIGNAL_MQTT_BINARY_MESSAGE_RECEIVED` or the like. We don't want to ever send binary data to a routine expecting a string.

2) Is there a way to configure for the topic whether we're expecting binary or utf-8 data? It would be preferable to declare this, otherwise we open ourselves to the potential of receiving binary data that happens to be decodable as utf-8, causing rare unexpected bugs.

pvizeli commented:
@armills i port mqtt to dispatcher on 0.39 so we can send everting throught it.

MrMep committed:
Update __init__.py

MrMep commented:
Ok,
I think I'm over my head here, with regards to HA.
Should I say something, I would say that we should able to manage binary payloads transparently and without any special handling, because there's no distinction at mqtt level between different type of payloads: afaik, there are no such things as payload "types", they are all binary actually.
I saw a message about dispatch being able to send around anything: I can confirm that no error arose after this modification. I'm receiving pictures from 5 different topics, and everything goes well...

Just one question about testing: there's a check in `tests/components/mqtt/test_init.py` exactly against the reception of binary payloads, I have no clue as to how manage the errors that comes out from that... 

thanks!

armills commented:
I think you're on the right track here, we just need to make sure additional safeguards are in place.

The problem is this: Even though mqtt doesn't care what the payload is, all code that subscribes to the event `SIGNAL_MQTT_MESSAGE_RECEIVED` is designed to accept a decoded string. If we want to pass binary data in hass, we need a new event to pass it.

MrMep commented:
Ok,
that looks like a huge mod: first thing that comes to my mind, we would need to change `async_subscribe`, to explicitly allow binary payloads.
I'm afraid I don't know HA well enough to consider all the implications...

MrMep commented:
Ok, after reflecting a little bit, I think this is not the place where to check the type of an mqtt payload. MQTT is data-agnostic and so HA's mqtt core should be, shouldn't it?

I suggest each component should be responsible for type control, and manage errors accordingly to its expectations (just adding a try-except would be enough for most cases, I guess).
I think it would be a more logical and a lighter move on the source compared with touching async_subscribe, create two separate paths for string and binary payloads, and the rest.

There aren't that many components (I count 8 in total) who receive mqtt messages: I tried sending a binary payload to:
- switches
- sensors
- binary sensors
- owntracks

sensor and owntrack throw an error in the log, but surely HA doesn't stop working.
After all, something must go very wrong for a binary payload to end up in a sensor or a switch topic...

Conversely, for example the local file camera component doesn't validate the content of the file, it just sends it to the front-end: if the file doesn't contain a picture, it just isn't rendered and an error is thrown...

balloob commented:
So sending bytes over MQTT comes up very rarely. And so whatever we do, we should make it backwards compatible as there is a lot of MQTT in use.

So, the easiest approach here is to not fix Home Assistant. Instead, just have the sender base64 encode the bytes and voila: you can send UTF-8.

If you really want to dig into all the internals of Home Assistant, there is a way to do it with minimal impact. In Home Assistant people subscribe to topics via `async_subscribe`. So if  that method is encoding aware, it's all good:

 - Add an encoding parameter to [`async_subscribe`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mqtt/__init__.py#L203-L218). Default to 'utf-8'. If topic matches, decode the bytes or don't do anything if encoding=None.
 - Update [`_mqtt_on_message`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mqtt/__init__.py#L565-L578) to not do anything with encoding and just forward it to dispatcher.

Now in your case you would just subscribe with encoding=None and get bytes.

MrMep committed:
Update __init__.py

MrMep committed:
Update __init__.py

MrMep commented:
@balloob I love your solution: easy, elegant and working.
This should be it,
thanks!
gl

balloob commented:
Okay, the code looks good. Now you need to make sure the tests work. Tests use this method: https://github.com/home-assistant/home-assistant/blob/dev/tests/common.py#L171-L175

Update that method to send bytes instead of string.

You will also need to write a test to verify that we can receive bytes.

MrMep committed:
git test - ignore

MrMep commented:
I'm sorry, I need some clarification and help.
First of all, we moved the utf-8 decoding from `on_message` to `async_subscribe`. Now, it seems to me that all the involved  tests run at dispatcher level (including the method @balloob asked me to modify): at that level we do not have the information about the nature of payload anymore.
This would mean that:
[test_receiving_non_utf8_message_gets_logged](https://github.com/home-assistant/home-assistant/blob/dev/tests/components/mqtt/test_init.py#L301-L331)
would have to go, or it should actually become a test to verify that we _can_ receive bytes...
Otherwise, I wouldn't know how to change the test to make it work, sorry.

Any help?
thanks
gl

MrMep committed:
Should work

MrMep committed:
minor fixes

MrMep commented:
Sorry guys, it took me a little bit to setup a local tox environment (mac os, there are a few glitches in the setup script) and to understand what you were talking about... Then I made a mess with git... I hope now everything is working...

So,
I changed the tests according to your directions, @balloob, and transformed the previous `test_receiving_non_utf8_message_gets_logged` into https://github.com/home-assistant/home-assistant/pull/6976/files#diff-797e3229a0d16bf1ddf33ccdcbcb6c3aR302
to test that binary messages get through.

I had to change `test_receiving_mqtt_message_fires_hass_event ` too, because now we're sending bytes for testing, not strings.

In the meanwhile, I added support for binary payloads to be published through the MQTT service: I added a configuration option `payload_file_path` that, when given and if the file is existing, gets sent as binary payload. If that's ok with you, I'll update the manual page accordingly (it needs some more configuration example in any case...)

thanks
gl

MrMep commented:
I would like to add that Zanzito, my android app, is fully working with the new changes: it exchanges pictures back and forth directly with HA now. It only needs the new mqtt camera component I wrote and that I will pull here as soon as this PR is approved.
If you want to try it, write me a PM and I'll send you a code to unlock it.

MrMep committed:
updated mqtt/services.yaml

armills commented:
Looks really good! This is a great solution. Just a few notes here:

1) Let's also add the `encoding` parameter to [`subscribe`](https://github.com/MrMep/home-assistant/blob/d7bf144cf9315562224f9ae4b78b877e6d657f11/homeassistant/components/mqtt/__init__.py#L238) for threaded components. It only needs to pass it on to `async_subscribe`.
2) We need to add tests for the new functionality in `async_subscribe`. You can see where we're now missing coverage [here](https://coveralls.io/builds/11012410/source?filename=homeassistant%2Fcomponents%2Fmqtt%2F__init__.py#L206). If you start with [this](https://github.com/MrMep/home-assistant/blob/d7bf144cf9315562224f9ae4b78b877e6d657f11/tests/components/mqtt/test_init.py#L134-L143) as your template, you can create two new tests. One that sends binary data to a utf-8 subscriber, and will mimic the [old test](https://github.com/home-assistant/home-assistant/pull/6976/files#diff-797e3229a0d16bf1ddf33ccdcbcb6c3aL322) with assertLogs. The second should just test sending binary data to a binary subscriber and ensure it's called successfully.

MrMep commented:
@armills, all clear! I had missed the threaded subscription.
Ok for the new tests, but isn't [this](https://github.com/home-assistant/home-assistant/pull/6976/files#diff-797e3229a0d16bf1ddf33ccdcbcb6c3aR302) already covering the second case? Or, at this point, my `test_receiving_binary_payload ` might be tautologic without calling `async_subscribe`... What do you think?

armills commented:
Good point. I think the second test case would replace the `test_receiving_binary_payload` test that currently exists in the PR. Since `_mqtt_on_message` is now unaware of encoding it no longer needs to be tested directly. We're basically moving the test to where we now perform the branching.

MrMep committed:
added two tests, modified threaded subscribe

MrMep commented:
@armills, I did what you asked for and updated the documentation

emlove reviewed (dismissed):
This LGTM. Great work!

armills commented:
Only problem is that somehow polymer got added to the PR: https://github.com/home-assistant/home-assistant/pull/6976/files#diff-d3eb3a918c38d80bb0a48ed78d8488ef

This is how to fix it, assuming your hass remote is named upstream:
```bash
git checkout upstream/dev -- homeassistant/components/frontend/www_static/home-assistant-polymer/
git submodule update
git commit
```

MrMep committed:
removing polymer

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -370,18 +389,30 @@ def async_publish_service(call):
+                    "'%s' failed because %s",
+                    msg_topic, payload_template, exc)
+                return
+        elif payload_file_path is not None:
```
I don't like this functionality. It's a big security leak. All of a sudden when you have the Home Assistant password we can send any file on the file system that the user has access to! Please remove.

MrMep replied:
Mmh, Isn't this exactly what the telegram notify platform does?
And actually you would need access to the local configuration files, not just the frontend, which is exactly like having access to the same files the platform has, or am I missing something?

balloob replied:
No, that is not ok either.  This is not a notification service, this is an MQTT implementation. It's up to platforms or components built on top to decide how to get bytes to send over MQTT. Please remove this feature.

MrMep replied:
So please advice how a component could send a binary payload if the underlying MQTT implementation does not allow for it. The way I see it, HA MQTT core implementation  does not conform to MQTT specifications [here](http://public.dhe.ibm.com/software/dw/webservices/ws-mqtt/mqtt-v3r1.html) and [here](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html):  

> A messaging transport that is agnostic to the content of the payload.

base64 encoding is a workaround and, if that is the answer, this whole PR is quite unnecessary, because it doesn't make HA compliant to the standard, which was my original goal.
Thanks

pvizeli replied:
I think you don't understand Paulus right. You should remove the sendfile function. Not more and not less.

MQTT is a base component. So If you want send files you can make a overlay component they support it. That could also be a custom component. I think also sendfile is not mqtt standard per default.

MrMep replied:
@pvizeli I think I understood @balloob just fine. What I don't understand is the lack of acknowledgement of the underlying problem and of some suggestion for alternate solutions. To be blunt, the underlying problem is that the base implementation of MQTT in HA was not protocol compliant. Can we please find a solution?

Given the very limited understanding I have of HA, I humbly offer a few thoughts:
It looks to me that there's no way, at the moment, a component could load a file or, for what is worth noting, build a binary chunk of data and publish it on an MQTT topic _as it is_ through the MQTT service. We are forced to put it into a JSON formatted string which, obviously, cannot contain binary data. Am I wrong?

Now, if this is a base, transport component, why offer a service at all? Shouldn't we have a separate component for that, one that manages _any_ kind of payload, as per mqtt specs?

IMHO, this ambiguity is well expressed, involuntary, in [this page](https://home-assistant.io/components/notify.mqtt/). I'm sure there are historical and practical reasons for this situation but, humbly, I dissent: this doesn't look like a good implementation to me.

So, you seem to be saying, I should write a new component that does exactly what the MQTT service does, except that it is not a service and it accepts binary payloads. Is that it?

thanks,
gl

pvizeli replied:
Paulus speak about recevie binary payloads. And that is supported with you change. So sending binary data is a other Topic.

First a service is a technic from hass and that is limited. So you can not send binary data trought a service call. It is also not the idea to use base64 for that. It is also not the idea that the mqtt service have full mqtt compatibility while it is a Service and give support to use it with scripts and automation they also not support binarys.

For binary stuff und IPC we have make dispatcher they allow to send data trought hass without passing trought the event bus. 

So we have 2 variants to do binary stuff on our components:
- the component access to hass.data[mqtt] object and send his data (my favor)
- we make a general dispatcher on mqtt component they support binary

I think you PR should first support to receive binary. For sending stuff, do that on your i.e. notify or what ever component.

MrMep replied:
Thanks Pascal, that's clear to me. I was just pointing out that I was building on top of a situation I hadn't created myself: the service is already there and it still looks incomplete and ambiguous to me, but ok, be it the way it is.

Just a quick clarification: so from a component the call would be
`hass.data['mqtt'].async_publish(msg_topic, payload, qos, retain)`
?

balloob replied:
No, from a component you always call `mqtt.async_publish(hass, …)` with `mqtt` being the MQTT component. One should not use `hass.data` entries from another component.

I am not sure where the concerns come from being spec compliant or not when we are only here talking about if we should be able to pass a file path in. Allowing HASS to expose any file in the system the running user has access to is a security issue and will not be part of a core component.

MrMep replied:
In fact, we were not talking about that. Anyway, thanks for the merge.

balloob replied:
This is a thread based on a comment on a line in a file which is exactly about being able to pass a file path in 🤔  So if this was no longer that, I am confused.

pvizeli replied:
Correct, we could add a function inside component to access into self for send binary payloads. With service call it will be not possible

MrMep replied:
Paulus, as I explained before, I was building on top of an existing service that doesn't allow binary payloads.
I tried to add a feature that is provided by the MQTT protocol but not by the implementation of this service. Admittedly, my first attempt was flawed, I hadn't thought about the security considerations. But after? How sending a file that is already available to the frontend (being in the `www` dir) could be possibly considered a security flaw? I even put in a path traversal safeguard...
And the argument that this is not something to be put in a base component, well, that is valid for the whole service, isn't it? I say this because there's no class "A" payload (strings) and class "B" payloads (the rest). They are all the same, actually they are all binary as you know,  and so they should be treated.

So the problem is not passing a file, but not being able to send binary data to the `publish` service and have them published _as is_

I reckon you disagree, well, no problem: I'll write an mqtt notifier component ;)

MrMep replied:
I'm deeply sorry @balloob and @pvizeli to come back here, just tell me if I have to move this question somewhere else more appropriate.

So, I'm working on a notifier component that sends images through MQTT. `mqtt.async_publish` works fine and the image is sent and received properly.

I have two problems: the first one is that the image data gets INFO logged, I guess by [this](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/core.py#L388), obviously spamming the log.

The second one is that the Recorder throws an error:
`ERROR (Recorder) [homeassistant.components.recorder.util] Error executing query: bytearray(THE WHOLE PICTURE CONTENT, AGAIN) is not JSON serializable`

I have no idea how to solve this, it's too hard 'core' ;)

Could you please suggest me how to move from here?

Thanks!
gl

balloob replied:
All data sent over the Hass event bus has to be JSON serializable.

MrMep replied:
Ok, thanks.
So what can we do? As far as I understood, if I call `mqtt. async_publish `, the payload is sent over the Hass event bus, but a `binaryarray` cannot be serialized, thus the error.
How can I circumvent that?

MrMep replied:
Never mind, I've found a solution. We'll talk about it with my PR.

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -201,15 +203,30 @@ def publish_template(hass, topic, payload_template, qos=None, retain=None):
+        if encoding is not None:
+            try:
+                payload = dp_payload.decode(encoding)
+                _LOGGER.info("Received message on %s: %s",
```
Please don't log this as info but as `debug` instead to avoid spamming the log.

MrMep replied:
It already was logged as info in _mqtt_on_message, but ok!

balloob commented on tests/common.py:
```diff
@@ -171,8 +171,11 @@ def mock_service_log(call):  # pylint: disable=unnecessary-lambda
 def async_fire_mqtt_message(hass, topic, payload, qos=0):
     """Fire the MQTT message."""
+    if isinstance(payload, str):
+        payload = bytearray(payload, 'utf-8')
```
Please use `payload.encode('utf-8')`. That returns a `bytes` object which is immutable.

MrMep replied:
done

MrMep committed:
requested changes

MrMep committed:
requested changes - minor fix

MrMep committed:
security wrap around payload_file_path

MrMep committed:
services.yaml updated

MrMep committed:
removed file publishing

MrMep committed:
minor fix

MrMep commented:
requested changes are done.
